FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc

```diff
@@ -0,0 +1,298 @@
+    }
+    // or create new SimTrack
+    else
+    {
+        simTrackIndex = simTracks_->size();
+        simTracks_->emplace_back(particle->pdgId(),
+                    particle->momentum(),
+                    particle->simVertexIndex(),
+                    particle->genParticleIndex());
+        simTracks_->back().setTrackId(simTrackIndex);
+    }
+    return simTrackIndex;
+}
+
+std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle()
+{
+    // only consider particles that start in the beam pipe and end outside the beam pipe
+    // try to get the decay time from pythia
+    // use hepmc units
+    // make the link simtrack to simvertex
+    // try not to change the simvertex structure
+    
+    // loop over gen particles
+    for ( ; genParticleIterator_ != genParticleEnd_ ; ++genParticleIterator_,++genParticleIndex_ ) 
+    {
+        // some handy pointers and references
+        const HepMC::GenParticle & particle = **genParticleIterator_;
+        const HepMC::GenVertex * productionVertex = particle.production_vertex();
+        const HepMC::GenVertex * endVertex = particle.end_vertex();
+
+        // skip incoming particles
+        if(!productionVertex){
+            continue;
+        }
+
+        // particle must be produced within the beampipe
+        if(productionVertex->position().perp2()*lengthUnitConversionFactor2_ > beamPipeRadius2_)
+        {
+            continue;
+        }
+        
+        // particle must not decay before it reaches the beam pipe
+        if(endVertex && endVertex->position().perp2()*lengthUnitConversionFactor2_ < beamPipeRadius2_)
+        {
+            continue;
+        }
+
+        // make the particle
+        std::unique_ptr<Particle> newParticle(
+            new Particle(particle.pdg_id(),
+                 math::XYZTLorentzVector(productionVertex->position().x()*lengthUnitConversionFactor_,
+                             productionVertex->position().y()*lengthUnitConversionFactor_,
+                             productionVertex->position().z()*lengthUnitConversionFactor_,
+                             productionVertex->position().t()*timeUnitConversionFactor_),
+                 math::XYZTLorentzVector(particle.momentum().x()*momentumUnitConversionFactor_,
+                             particle.momentum().y()*momentumUnitConversionFactor_,
+                             particle.momentum().z()*momentumUnitConversionFactor_,
+                             particle.momentum().e()*momentumUnitConversionFactor_)));
+        newParticle->setGenParticleIndex(genParticleIndex_);
+
+        // try to get the life time of the particle from the genEvent
+        if(endVertex)
+        {
+            double labFrameLifeTime = (endVertex->position().t() - productionVertex->position().t())*timeUnitConversionFactor_;
+            newParticle->setRemainingProperLifeTimeC(labFrameLifeTime / newParticle->gamma() * fastsim::Constants::speedOfLight);
```

Comment:
shouldn't this be 
labFrameLifeTime * fastsim::Constants::speedOfLight / newParticle->gamma()
Otherwise it seems we are dividing tau by c instead of multiplying them?

Reply:
Oh, that's the same thing; gamma is the only thing in the denominator in both cases since I did not use any brackets. I can swap the order of gamma and c if you think that's easier to read, though.

Reply:
But the line has a multiplication by `fastsim::Constants::speedOfLight`, no?

(of course the order of computations can be discussed for other reasons)

Reply:
You are right.  This is fine.  Somehow thought the gamma went to the denominator, but it doesn't.